File: dailyQuestion/2020-05/05-18/golang/solution.go
package main

import (
	"fmt"
	"math"
)

func main() {
	a := []int{2, 3, -2, 4}

	b := maxProduct(a)

	fmt.Println(b)
}

// 动态规划
func maxProduct(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	nums_min := nums[0]
	nums_max := nums[0]
	result := nums[0]
	for i := 1; i < len(nums); i++ {
		a := nums_min
		b := nums_max
		nums_min = min(nums[i], nums[i]*a, nums[i]*b)
		nums_max = max(nums[i], nums[i]*a, nums[i]*b)
		if nums_max > result {
			result = nums_max
		}
	}
	return result
}

func max(a, b, c int) int {
	if a > b {
		if a > c {
			return a
		}
		return c
	}
	if b > c {
		return b
	}
	return c
}

func min(a, b, c int) int {
	if a > b {
		if c > b {
			return b
		}
		return c
	}
	if a > c {
		return c
	}
	return a
}

// 暴力求解
func maxProduct2(nums []int) int {
	result := math.MinInt64
	for i := 0; i < len(nums); i++ {
		tmp := 1
		for j := i; j >= 0; j-- {
			tmp *= nums[j]
			if tmp > result {
				result = tmp
			}
		}
	}
	return result
}

File: dailyQuestion/2020-05/05-18/python/solution.py
# 动态规划求解
def maxProduct(nums: [int]) -> int:
    max_nums = nums[0]
    min_nums = nums[0]
    result = nums[0]
    for i in range(1, len(nums)):
        tmp_max = max_nums
        tmp_min = min_nums
        max_nums = max(nums[i], nums[i] * tmp_max, nums[i] * tmp_min)
        min_nums = min(nums[i], nums[i] * tmp_max, nums[i] * tmp_min)
        if max_nums > result:
            result = max_nums
    return result


# 暴力求解
def maxProduct2(nums: [int]) -> int:
    result = float('-inf')
    for i in range(len(nums)):
        tmp = 1
        for j in range(i, -1, -1):
            tmp *= nums[j]
            if tmp > result:
                result = tmp
    return result


if __name__ == "__main__" :
    nums = [2,0,3,-2,3,4,4,5,5]
    result = maxProduct(nums)
    print(result)